ocaml/xapi/xapi_host_driver.ml

```diff
@@ -0,0 +1,297 @@
+(*
+   Copyright (c) Cloud Software Group, Inc.
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU Lesser General Public License as published
+   by the Free Software Foundation; version 2.1 only. with the special
+   exception on linking described in file LICENSE.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU Lesser General Public License for more details.
+ *)
+
+module D = Debug.Make (struct let name = __MODULE__ end)
+
+open D
+module Unixext = Xapi_stdext_unix.Unixext
+module DriverMap = Map.Make (String)
+module DriverSet = Set.Make (String)
+
+let ( // ) = Filename.concat
+
+let invalid_value field value =
+  raise Api_errors.(Server_error (invalid_value, [field; value]))
+
+let internal_error fmt =
+  Printf.ksprintf
+    (fun msg ->
+      error "%s" msg ;
+      raise Api_errors.(Server_error (internal_error, [msg]))
+    )
+    fmt
+
+let create ~__context ~host ~name_label ~versions ~active_version
+    ~selected_version =
+  info "%s: %s" __FUNCTION__ name_label ;
+  let ref = Ref.make () in
+  let uuid = Uuidx.to_string (Uuidx.make ()) in
+  Db.Host_driver.create ~__context ~ref ~uuid ~host ~name_label ~versions
+    ~active_version ~selected_version
+
+let destroy ~__context ~self =
+  D.debug "Destroying driver %s" (Ref.string_of self) ;
+  Db.Host_driver.destroy ~__context ~self
+
+(*
+* `/lib/modules/<kernel-version>/updates/` is searched by the kernel for
+  drivers - this is `selected_dir`.
+* The hierarchy is expected to contain symbolic links to the file
+  actually containing the driver in
+  `/lib/modules/<kernel-version>/xenserver/<driver>/<version>/<name>.ko` -
+  the base directory here is `all_versions_dir`. *)
+type driver_paths = {selected_dir: string; all_versions_dir: string}
+
+let get_drivers_base_path ~__context ~host =
+  let linux_version =
+    List.assoc "linux" (Db.Host.get_software_version ~__context ~self:host)
+  in
+  let base_dir = "/lib/modules" // linux_version in
+  {
+    selected_dir= base_dir // "updates"
+  ; all_versions_dir= base_dir // "xenserver"
+  }
+
+let reload_driver_selection ~__context ~host =
+  let linux_version =
+    List.assoc "linux" (Db.Host.get_software_version ~__context ~self:host)
+  in
+  try
+    let _, _ =
+      Forkhelpers.execute_command_get_output "/usr/sbin/depmod"
```

Comment:
I'm not sure if this `depmod` could load the driver module into the running Kernel or not.
Do we need to reboot the host to get the module loaded into the kernel actually, or use `modprobe`?

Reply:
We don't necessarily need to reboot, you're right. `modprobe` operates on a single driver, we will need to trigger a udev event for device drivers to be added again instead.

Reply:
Thanks. Now I refresh my mind as the following. Please correct me if I'm wrong.
1. The device is already equipped on the host, when doing `select`,  the symbol link under `/lib/modules/<kernel-version>/updates/` is (re)created, and the udevadm command can trigger the Kernel to reload the new driver for the device.
2. The device is not equipped on the host yet, when doing `select`, the symbol link is (re)created, but no need to load it since it will be automatically when the device is present. But udevadm is harmless in this case.

Reply:
Precisely!